server/lib/flow-event.js

```diff
@@ -2,28 +2,124 @@
-var os = require('os');
-var Promise = require('bluebird');
+const _ = require('lodash');
+const config = require('./configuration');
+const flowMetrics = require('./flow-metrics');
+const os = require('os');
 
-var DNT_ALLOWED_DATA = [
+const DNT_ALLOWED_DATA = [
   'context',
   'entrypoint',
   'migration',
   'service',
 ];
-var NO_DNT_ALLOWED_DATA = DNT_ALLOWED_DATA.concat([
+const NO_DNT_ALLOWED_DATA = DNT_ALLOWED_DATA.concat([
   'utm_campaign',
   'utm_content',
   'utm_medium',
   'utm_source',
   'utm_term'
 ]);
-var HOSTNAME = os.hostname();
-var MAX_DATA_LENGTH = 100;
-var VERSION = 1;
+const HOSTNAME = os.hostname();
+const MAX_DATA_LENGTH = 100;
+const VERSION = 1;
 
-module.exports = function (event, data, request) {
+const FLOW_BEGIN_EVENT_TYPES = /^flow\.[a-z_-]+\.begin$/;
+const FLOW_ID_KEY = config.get('flow_id_key');
+const FLOW_ID_EXPIRY = config.get('flow_id_expiry');
+
+const ENTRYPOINT_PATTERN = /^[\w\.-]+$/;
+const SERVICE_PATTERN = /^(sync|content-server|none|[0-9a-f]{16})$/;
+const VALID_FLOW_EVENT_PROPERTIES = [
+  { key: 'client_id', pattern: SERVICE_PATTERN },
+  { key: 'context', pattern: /^[0-9a-z_-]+$/ },
+  { key: 'entryPoint', pattern: ENTRYPOINT_PATTERN },
+  { key: 'entrypoint', pattern: ENTRYPOINT_PATTERN },
+  { key: 'flowId', pattern: /^[0-9a-f]{64}$/ },
+  { key: 'migration', pattern: /^(sync11|amo|none)$/ },
+  { key: 'service', pattern: SERVICE_PATTERN }
+];
+
+const IS_DISABLED = config.get('client_metrics').stderr_collector_disabled;
+
+module.exports = (req, metrics, requestReceivedTime) => {
+  if (IS_DISABLED || ! isValidFlowData(metrics, requestReceivedTime)) {
+    return;
+  }
+
+  const events = metrics.events || [];
+  const flowEvents = _.filter(events, event => {
+    return event.type.indexOf('flow.') === 0;
+  });
+
+  flowEvents.forEach(event => {
+    if (FLOW_BEGIN_EVENT_TYPES.test(event.type)) {
+      event.time = metrics.flowBeginTime;
+      event.flowTime = 0;
+    } else {
+      event.time = estimateTime({
+        /*eslint-disable sorting/sort-object-props*/
+        start: metrics.startTime,
+        offset: event.offset,
+        sent: metrics.flushTime,
+        received: requestReceivedTime
+        /*eslint-enable sorting/sort-object-props*/
+      });
+
+      if (! isValidTime(event.time, requestReceivedTime)) {
+        return;
+      }
+
+      event.flowTime = event.time - metrics.flowBeginTime;
+    }
+
+    logFlowEvent(event, metrics, req);
+  });
+};
+
+function isValidFlowData (metrics, requestReceivedTime) {
+  if (! metrics.flowId) {
+    return false;
+  }
+
+  if (! isValidTime(metrics.flowBeginTime, requestReceivedTime)) {
+    return false;
+  }
+
+  if (! VALID_FLOW_EVENT_PROPERTIES.every(p => isValidProperty(metrics[p.key], p.pattern))) {
+    return false;
+  }
+
+  return flowMetrics.validate(FLOW_ID_KEY, metrics.flowId, metrics.flowBeginTime, metrics.agent);
+}
+
+function isValidTime (time, requestReceivedTime) {
+  if (typeof time !== 'number') {
```

Comment:
I could use `_.isNumber` here but, for the record, I'm not a fan.

Fully unwound, the implementation of `_.isNumber` in current lodash master looks like this:

```js
var numberTag = '[object Number]';
var nullTag = '[object Null]';
var undefinedTag = '[object Undefined]';

var symToStringTag = Symbol ? Symbol.toStringTag : undefined;

var objectProto = Object.prototype;
var hasOwnProperty = objectProto.hasOwnProperty;
var nativeObjectToString = objectProto.toString;

function isNumber(value) {
  return typeof value == 'number' ||
    (isObjectLike(value) && baseGetTag(value) == numberTag);
}

function isObjectLike(value) {
  return value != null && typeof value == 'object';
}

function baseGetTag(value) {
  if (value == null) {
    return value === undefined ? undefinedTag : nullTag;
  }
  value = Object(value);
  return (symToStringTag && symToStringTag in value)
    ? getRawTag(value)
    : objectToString(value);
}

function getRawTag(value) {
  var isOwn = hasOwnProperty.call(value, symToStringTag),
      tag = value[symToStringTag];

  try {
    value[symToStringTag] = undefined;
    var unmasked = true;
  } catch (e) {}

  var result = nativeObjectToString.call(value);
  if (unmasked) {
    if (isOwn) {
      value[symToStringTag] = tag;
    } else {
      delete value[symToStringTag];
    }
  }
  return result;
}

function objectToString(value) {
  return nativeObjectToString.call(value);
}
```

It contains all that extra complexity to handle cases we're not interested in. We already know that our code is not passing us the result of `new Number(time)` or anything else funky like that. Literally all we want is the very first condition that lodash checks, `typeof time === 'number'`. If that test fails we know straight away that `time` is invalid and can immediately `return false`.

Fair enough, in the common case we expect that test to pass, so the logic will short-circuit the remainder of lodash's implementation. But in the uncommon, failing case, it is going to do a whole load of testing that we're not interested in and we know will fail anyway. What's the point?